File: masar_gig/custom/material_request/material_request.js
frappe.ui.form.on('Material Request', {
    refresh: function(frm) {
        if (frm.doc.__islocal == 1) {
            setDefaultRequester(frm);
        }

        // if (frm.doc.docstatus == 1) {
            addSQItems(frm);
        // }
    },
    onload: function(frm) {
        if (frm.doc.__islocal == 1) {
            setDefaultRequester(frm);
        }

        // if (frm.doc.docstatus == 1) {
            addSQItems(frm);
        // }
    }
  });

  function setDefaultRequester(frm) {
    frappe.call({
        method: "masar_gig.custom.material_request.material_request.get_emp",
        args: {
            user: frappe.session.user
        },
        callback: function(r) {
            if (r.message) {
                console.log(r.message);
                var emp_name = r.message;
                console.log("Test 1", emp_name);
                frappe.model.set_value(frm.doc.doctype, frm.doc.name, "custom_requester", emp_name);
                frm.refresh_field("custom_requester");
                console.log("Test 2", frm.doc.custom_requester);
            }
        }
    });
  }


  function addSQItems(frm) {
    if (!frm.doc.custom_prices_inquiry && frm.doc.docstatus == 0) {
        frm.add_custom_button(__('Get Items from Supplier Quotation'), function () {
            frappe.call({
                method: 'masar_gig.custom.material_request.material_request.get_supplier_quotation',
                args: { mr_name: frm.doc.name },
                callback: function (r) {
                    if (r.message && r.message.length) {
                        let dialog = new frappe.ui.Dialog({
                            title: __('Select Supplier Quotation'),
                            fields: [
                                {
                                    fieldtype: 'Link',
                                    fieldname: 'sq_name',
                                    label: __('Supplier Quotation'),
                                    options: 'Supplier Quotation',
                                    reqd: 1,
                                    get_query: function() {
                                        return {
                                            filters: { docstatus: 1 }
                                        };
                                    }
                                }
                            ],
                            primary_action_label: __('Select The Supplier Quotation'),
                            primary_action: function () {
                                let selected_data = dialog.get_values();
                                if (!selected_data || !selected_data.sq_name) {
                                    frappe.msgprint(__('Please select a Supplier Quotation.'));
                                    return;
                                }

                                frappe.call({
                                    method: "masar_gig.custom.material_request.material_request.get_sq_items",
                                    args: { sq: selected_data.sq_name },
                                    callback: function (r) {
                                        if (r.message && r.message.length) {
                                            frm.clear_table("items");
                                            r.message.forEach(item => {
                                                let child = frm.add_child("items");
                                                Object.assign(child, item);
                                            });
                                            frm.refresh_field("items");
                                            dialog.hide();
                                            frm.save();
                                        } else {
                                            frappe.msgprint(__('No items found in the selected Supplier Quotation.'));
                                        }
                                    }
                                });
                            }
                        });

                        dialog.show();
                    } else {
                        frappe.msgprint(__('No Supplier Quotations available.'));
                    }
                }
            });
        });
    }
}

File: masar_gig/custom/material_request/material_request.py
import frappe
from frappe import _

def validate(self, method):
    calc_total(self)

@frappe.whitelist()
def get_emp(user):
    emp_sql = frappe.db.sql("""
            SELECT name FROM tabEmployee WHERE user_id = %s
        """, (user), as_dict=True)
    
    if emp_sql and emp_sql[0] and emp_sql[0]['name']:
        return emp_sql[0]['name']
    
@frappe.whitelist()
def get_supplier_quotation(mr_name):
    if mr_name:
        sql = frappe.db.sql("""
                SELECT DISTINCT tsq.name 
                FROM `tabSupplier Quotation` tsq 
                INNER JOIN `tabSupplier Quotation Item` tsqi ON tsqi.parent = tsq.name
                INNER JOIN `tabMaterial Request` tmr ON tsqi.material_request = tmr.name
                WHERE tmr.docstatus = 1 and tsq.docstatus = 1
            """, as_dict=True)
        
        if sql:
            return sql

@frappe.whitelist()
def get_sq_items(sq):
    sq_doc = frappe.get_doc("Supplier Quotation", sq)
    
    item_rates = {item.item_code: {"rate": item.rate, "qty": item.qty, "amount": item.amount, "item_name": item.item_name} 
                  for item in sq_doc.items}
    
    items = []
    for item in sq_doc.items:
        item_dict = {
            "item_code": item.item_code,
            "uom": item.uom,
            "conversion_factor": item.conversion_factor,
            "qty": item.qty,
            "rate": item.rate,
            "item_name": item.item_name,
            "amount": item.amount,
            "custom_supplier_quotation": sq,
            "custom_supplier_quotation_item": item.name
        }
        items.append(item_dict)
    
    return items

def calc_total(self):
    total = 0
    if self.items:
        for item in self.items:
            if hasattr(item, 'custom_supplier_quotation') and item.custom_supplier_quotation:
                sq_item = frappe.db.get_value(
                    "Supplier Quotation Item",
                    item.custom_supplier_quotation_item,
                    ["rate", "amount"],
                    as_dict=True
                )
                if sq_item:
                    item.rate = sq_item.rate
                    item.amount = item.qty * item.rate
            
            if item.amount:
                total += item.amount
                
    self.custom_total = total

def on_update(self, method):
    update_rates_from_sq(self)

def update_rates_from_sq(self):
    """Update rates from supplier quotation after save"""
    if self.items:
        for item in self.items:
            if hasattr(item, 'custom_supplier_quotation') and item.custom_supplier_quotation:
                sq_item = frappe.db.get_value(
                    "Supplier Quotation Item",
                    item.custom_supplier_quotation_item,
                    ["rate"],
                    as_dict=True
                )
                if sq_item:
                    frappe.db.set_value(
                        "Material Request Item",
                        item.name,
                        {
                            "rate": sq_item.rate,
                            "amount": item.qty * sq_item.rate
                        }
                    )
        